### PR TITLE
Add support for arbitrary predicates for multi cursor queues

### PR DESCRIPTION
**Pull request by bergundy**
## What changed?

- Add support for arbitrary predicates for multi cursor queues, adds a Grouper abstraction for grouping tasks and constructing predicates
- Add a DestinationPredicate (will be used later in multi-destination queues)
- Add Grouper implementations for namespace ID and namespace ID + destination.

## Why?

We need this flexibility for Nexus, this is some of the prep work.

## How did you test it?

Add some unit tests, ran existing tests.

## Potential risks

We should look closely at this as it may mess up task processing multi-cursor logic.

## Is hotfix candidate?

No.

### CODE REVIEW

File: service/history/queues/grouper.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+// Grouper groups tasks and constructs predicates for those groups.
+type Grouper interface {
+	// Key returns the group key for a given task.
+	Key(task tasks.Task) (key any)
+	// Predicate constructs a prdicate from a slice of keys.
+	Predicate(keys []any) tasks.Predicate
+}
+
+type GrouperNamespaceID struct {
```
Comment (tdeebswihart) [inline service/history/queues/grouper.go:38]: nit: What about `GroupByNamespaceID` instead? It reads a bit better imo

File: service/history/queues/slice.go [hunk 2]
```diff
@@ -344,17 +347,15 @@ func (s *SliceImpl) shrinkPredicate() {
 		return
 	}
 
-	if len(s.executableTracker.pendingPerNamespace) > shrinkPredicateMaxPendingNamespaces {
+	// TODO: this should be generic enough to shrink any predicate type, probably doesn't belong here.
+	pendingPerKey := s.executableTracker.pendingPerKey
+	if len(pendingPerKey) > shrinkPredicateMaxPendingKeys {
 		// only shrink predicate if there're few namespaces left
```
Comment (yycptt) [inline service/history/queues/slice.go:353]:
```suggestion
		// only shrink predicate if there're few keys left
```

File: service/history/queues/slice.go [hunk 3]
```diff
@@ -344,17 +347,15 @@ func (s *SliceImpl) shrinkPredicate() {
...
 	}
 
-	pendingNamespaceIDs := make([]string, 0, len(s.executableTracker.pendingPerNamespace))
-	for namespaceID := range s.executableTracker.pendingPerNamespace {
-		pendingNamespaceIDs = append(pendingNamespaceIDs, namespaceID.String())
-	}
-	namespacePredicate := tasks.NewNamespacePredicate(pendingNamespaceIDs)
-	s.scope.Predicate = tasks.AndPredicates(s.scope.Predicate, namespacePredicate)
+	minimalPredicate := s.grouper.Predicate(maps.Keys(pendingPerKey))
+	s.scope.Predicate = tasks.AndPredicates(s.scope.Predicate, minimalPredicate)
```
Comment (yycptt) [inline service/history/queues/slice.go:358]: now that I think about it, the `tasks.AndPredicates` looks unnecessary to me.

**Review by yycptt — APPROVED** [review]

File: service/history/queues/slice.go (hunk 3 above)
Comment (bergundy) [reply service/history/queues/slice.go:358]: Yeah, good catch!

File: service/history/queues/queue_scheduled.go [hunk 4]
```diff
@@ -135,6 +135,7 @@ func NewScheduledQueue(
 			options,
 			hostRateLimiter,
 			readerCompletionFn,
+			GrouperNamespaceID{},
```
[the remark this answers is not in the document]
Comment (bergundy) [reply service/history/queues/queue_scheduled.go:138]: Let's wait for a use case, so far I don't see one.

**Pull request merged**